File: Sprint-1/JavaScript/hasPairWithSumSorted/hasPairWithSumSorted.js
/**
 * Find if there is a pair of numbers that sum to target (Sorted version)
 * Time Complexity:
 * Space Complexity:
 * @param {Array<number>} numbers - Array of numbers to search through
 * @param {number} targetSum - Target sum to find
 * @returns {boolean} True if pair exists, false otherwise
 */
export function hasPairWithSumSorted(numbers, targetSum) {
  // Create sorted copy to avoid modifying input
  const sortedNumbers = [...numbers].sort(
    (firstNumber, secondNumber) => firstNumber - secondNumber
  );

  for (
    let currentIndex = 0;
    currentIndex < sortedNumbers.length;
    currentIndex++
  ) {
    const requiredMatch = targetSum - sortedNumbers[currentIndex];

    for (
      let searchIndex = currentIndex + 1;
      searchIndex < sortedNumbers.length;
      searchIndex++
    ) {
      if (sortedNumbers[searchIndex] === requiredMatch) {
        return true;
      }
      // Early exit if we've gone too far
      if (sortedNumbers[searchIndex] > requiredMatch) {
        break;
      }
    }
  }
  return false;
}

File: Sprint-1/Python/has_pair_with_sum_sorted/has_pair_with_sum_sorted.py
"""
Find if there is a pair of numbers that sum to target. 
Optimise your solution with a precomputation.
Time Complexity: 
Space Complexity:
"""
from typing import List

def has_pair_with_sum_sorted(numbers: List[float], target_sum: float) -> bool:
    sorted_numbers = sorted(numbers)
    left_index = 0
    right_index = len(sorted_numbers) - 1
    
    while left_index < right_index:
        current_sum = sorted_numbers[left_index] + sorted_numbers[right_index]
        if current_sum == target_sum:
            return True
        if current_sum < target_sum:
            left_index += 1
        else:
            right_index -= 1
            
    return False 